pkgs/build-support/fetchurl/builder.sh

```diff
@@ -142,18 +142,20 @@ for url in "${urls[@]}"; do
             if test -n "${!varName}"; then mirrors="${!varName}"; fi
 
             for url3 in $mirrors; do
-                urls2="$urls2 $url3$fileName";
+                urlsResolved+=("$url3$fileName");
             done
         fi
     fi
 done
-urls="$urls2"
+urls=("${urlsResolved[@]}")
+
+unset urlsResolved
 
 # Restore globbing settings
 set +o noglob
 
 if test -n "$showURLs"; then
-    echo "$urls" > $out
+    echo "$urls[*]" > $out
```

Comment:
> Wait a minute. I found a syntax error in the seldom-checked `showURL` feature.

Ah good spot!

Reply:
BTW, I improved the test case to ensure that the resolved URL orders are also the same.

Reply:
> Can we add test coverage for `showURLs`?

The new test looks good, but I don't think it actually adds coverage for `showURLs = true`?

Not sure there's a simple way to capture the build log though, without (ab)using `testBuildFailure`?

Reply:
I tried, but the old behaviour specifically specify the `stdenv.mkDerivation`-`name` as `urls` no matter what `name` is passed to the fetcher., and `tests.invalidateFetcherByDrvHash` explicitly relies on passing an alternative `name` to the fetcher.

What makes the problem more complex is `fetchurl`'s `name` argument being considered the "default name" to fall back to when `pname` and `version` are not presented for the convenience of derived fetchers.

The one-off solution would be reimplementing the tester to use `overrideAttrs`. (To minimize the impact, we could do it only when the `name` argument change doesn't change the FOD's `drvPath`.) Another is to change how the `fetchurl` passes `name` when `showURL` is set as `true`.